text/0000-vec-fallible-allocation.md

```diff
@@ -0,0 +1,320 @@
+below) while we explore other alternatives. We have no plans to stabilize this feature in its current form and fully expect that these methods will be removed
+in the future.
+
+# Motivation
+[motivation]: #motivation
+
+The motivation for this change is well documented by other RFCs, such as the proposed [fallible_allocation feature](https://github.com/rust-lang/rfcs/pull/3140)
+and the accepted [fallible_collection_alloc feature](https://github.com/rust-lang/rfcs/blob/master/text/2116-alloc-me-maybe.md).
+
+As a brief summary, there are environments where dynamic allocation is required and failing to allocate is both expected and able to be handled. For example,
+in OS Kernel environments (such as [Linux](https://lore.kernel.org/lkml/CAHk-=wh_sNLoz84AUUzuqXEsYH35u=8HV3vK-jbRbJ_B-JjGrg@mail.gmail.com/)), some embedded
+systems, high-reliability systems (such as databases) and multi-user services (where a single request may fail without the entire service halting, such as a
+web server).
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+One often unconsidered failure in a software system is running out of memory (commonly referred to as "Out-of-memory" or OOM): if code attempts to dynamically
+allocate memory, and there is an insufficient amount of available memory on the system, how should this be handled? In most applications, the failure to
+allocate means that the code cannot make any progress, and so the appropriate response is to exit immediately indicating that there was a failure (i.e., to
+`panic`). Consider the Rust Compiler itself: if the compiler cannot allocate memory then it cannot fail that part of the compilation and attempt to continue
+the rest of the compilation - the only appropriate action is to fail the entire compilation and emit an error message. Since panicking on allocation failure is
+appropriate in most circumstances, this is the default within the Rust standard library. Confusingly, this approach is referred to as "infallible allocation":
+from the perspective of something calling a function it can assume that allocations cannot fail (otherwise the entire process will be terminated).
+
+There are, however, environments where allocation failures are both expected and able to be handled. Consider a multi-user service such as a database or a web
+server: if a request comes in to the service that requires allocating more memory than is available on the system, then the service can respond with a failure
+for just that request, and it can continue to service other requests. This approach is referred to as "fallible allocation": each function that may allocate
+needs to indicate to its caller if an attempt to allocate has failed. Within the Rust standard library, one can identify "fallible allocation" functions by
+their names being prefixed with `try_` and their return types being an instance of `Result`.
```

Comment:
Tried to break up into more paragraphs. I also like doing once sentence per line because it makes for nice diffs (after the switch!) Conversely, one can wrap at fixed length, have nice diffs, but not both.

```suggestion
One often unconsidered failure in a software system is running out of memory (commonly referred to as "Out-of memory" or OOM): 
if code attempts to dynamically allocate memory, and there is an insufficient amount of available memory on the system, how should this be handled?

In most applications, the failure to allocate means that the code cannot make any progress, and so an appropriate response is to exit immediately indicating that there was a failure (i.e., to `panic`).
Consider the Rust Compiler itself: if the compiler cannot allocate memory then it cannot fail that part of the compilation and attempt to continue the rest of the compilation --- the only appropriate action is to fail the entire compilation and emit an error message.

Panicking on allocation failureis the default within the Rust standard library for these reasons.
This approach is referred to as "infallible allocation", rather confusingly.
The name is from the perspective of something calling a function.
The caller can assume that allocations cannot fail, for if it indeed doesn't the assumption is valid, and if it in fact the does then the program panics and the caller will not live to observe the failure.

There are, however, environments where allocation failures are expected and also required to be handled more gracefully.
Consider a multi-user service such as a database or a web server:
if a request comes in to the service that requires allocating more memory than is available on the system, then the service can respond with a failure for just that request, and it can continue to service other requests.
This approach is referred to as "fallible allocation".
Each function that may allocate needs to indicate to its caller if an attempt to allocate has failed, and the caller needs to explicitly handle that possibility.

Within the Rust standard library, one can identify "fallible allocation" functions by their names being prefixed with `try_` and their return types being an instance of `Result`.
```

Reply:
I've split up the paragraphs and taken some of your wording (thanks!).
I wasn't sure if there was a standard for formatting markdown files, but splitting per-sentence to make the diffs nicer makes sense.

Reply:
Looks good! Feel free to mark this thread resolved.